### PR TITLE
Add a `fake` method to Cashier

### PR DESCRIPTION
**Pull request by kylemilloy**
# Overview
This PR adds a way for users to easily fake events and API calls with a single command and a couple options for customizing the responses and asserting that certain events are dispatched.

# How it works
Under the hood, `Cashier::fake()` creates a new instance of `CashierFake` which, upon creation, fakes API calls with `Http::fake()` and Cashier-provided events as well with `Event::fake()`. A user may customize this experience if they wish to override the defaults by providing an array with an endpoint as the key and a response as its value as such:

```php
Cashier::fake([
  'subscription/create' => ['fake-response' => 'and its data']
]);
```

Internally, CashierFake will format this path into a paddle API call within the `retrieveEndpoint($path)` call which uses Cashier's `vendorsUrl` method and some constants to return an endpoint such as: `vendors.paddle.com/api/2.0/subscription/create`. From here it will take the faked response and either return it as part of a `Http::response` if it's an array or allow the user to customize that behaviour by passing a function. 

There are two additional methods that can be used: `paypal()` and `card()`. These functions change the information returned by the `subscription/users` endpoint with `paypal()` sending a response that the user is paying via PayPal and `card()` that it is paying through a Visa with the following parameters:

```js
{
  card_type: 'visa',
  last_four_digits: '1234',
  expiry_date: '04/2022',
}
```

If these values need to be overwritten, you may pass your own information to the `card()` method and it will merge this into the response as such:

```php
Cashier::fake()->card(['card_type' => 'mastercard']);
```

Finally the `CashierFake` method also fakes Events provided by the library. These events are:

- Laravel\Paddle\Events\PaymentSucceeded
- Laravel\Paddle\Events\SubscriptionCancelled
- Laravel\Paddle\Events\SubscriptionCreated
- Laravel\Paddle\Events\SubscriptionPaymentFailed
- Laravel\Paddle\Events\SubscriptionPaymentSucceeded
- Laravel\Paddle\Events\SubscriptionUpdated

A user may append additional events to this process if they wish by adding passing them to the fake such as:

```php
Cashier::fake(events: CustomEvent::class);

// or
Cashier::fake(events: [CustomEvent::class, AndAnotherOne::class]);
```

This also provides a number of assertions, present both on the `CashierFake` and `Cashier` class. The assertion methods on the `Cashier` act as a proxy for the `CashierFake` and forward the requests into that class. I went with this approach instead of putting the assertions only on the `Cashier` as I felt importing all the events it needed to cover, plus the Event Facade felt messy and distracts a dev who is trying to understand the concerns of the `Cashier` class. Open to feedback on this one or if someone wants to recommend a better way of doing this.

Example of assertion syntax:
```php
// Switching between Cashier/CashierFake sucks
Cashier::fake()

CashierFake::assertSubscriptionCreated();

// This looks better:
Cashier::fake();

Cashier::assertSubscriptionCreated();

// However this looks like garbage...so again...placed the assertions in CashierFake and 
// did a bit of code duplication in favour of import hell.
use Illuminate\Support\Facades\Http;
use Laravel\Paddle\Exceptions\PaddleException;
use Money\Currencies\ISOCurrencies;
use Money\Currency;
use Money\Formatter\IntlMoneyFormatter;
use Money\Money;
use NumberFormatter;
use Illuminate\Support\Facades\Event;
use Laravel\Paddle\Events\PaymentSucceeded;
use Laravel\Paddle\Events\SubscriptionCancelled;
use Laravel\Paddle\Events\SubscriptionCreated;
use Laravel\Paddle\Events\SubscriptionPaymentFailed;
use Laravel\Paddle\Events\SubscriptionPaymentSucceeded;
use Laravel\Paddle\Events\SubscriptionUpdated;

class Cashier
{
  // ...
}
```

# One final thing...

Let me know if you want me to remove this but I found that setting `<server />` variables in my PHPUnit did nothing...`<env />` did...so I've updated the `CONTRIBUTING.md` file with that. I've also added an env there for `PADDLE_SANDBOX` because it's likely that you're going to be testing in sandbox mode...not sure if this should also be a part of the `phpunit.xml.dist` too...lemme know.

### CODE REVIEW

**Comment by kylemilloy** [comment]
Screen Shot 2022-05-13 at 15 59 57

Right off the bat I can see I'm going to have to change somethings to make this compatible...I'll make those changes over the weekend.

@driesvints - Let me know what you think...I'm kinda iffy about the `endpoints()` method in `CashierFaker`...just seems like a lot of boilerplate that might be better off elsewhere but you lemme know what you'd like to see.

**Comment by kylemilloy** [comment]
Nevermind...just removed the named arg and it worked fine. Ready for review @driesvints

File: src/CashierFake.php
Comment (kylemilloy) [inline src/CashierFake.php:207]: This method smells...would love some input to make this nice-er.

File: src/CashierFake.php
Comment (kylemilloy) [inline src/CashierFake.php:265]: I don't see Laravel leveraging consts very often...so not sure if you'd prefer this as something else or if there's a better naming convention you'd like.

File: .github/CONTRIBUTING.md
Comment (kylemilloy) [inline .github/CONTRIBUTING.md:11]: I noted this in the PR but just calling it out here...I found that adding `<server>` did nothing but `<env>` got the result of injecting the environment variable for the tests. I also added SANDBOX to this too because it's likely that you'll be running tests in the sandbox...lemme know if you want me to yank it.

File: src/CashierFake.php
Comment (kylemilloy) [reply src/CashierFake.php]: Removed these magic strings in favour of hardcoded. Internally they're called in one place and the idea is that a user could override them without having to know the path but then that raises questions of "do we include some endpoints and not others" and then the feel of it was janky too...doing `Cashier::faker(CashierFake::PATH_SUBSCRIPTION_REFUND)` with a mix of Cashier/CashierFake looked like hot trash.

File: src/CashierFake.php
Comment (driesvints) [inline src/CashierFake.php:246]: I don't think we should provide default responses and always let people indicate what should be returned.

File: tests/Feature/WebhooksTest.php
Comment (driesvints) [inline tests/Feature/WebhooksTest.php:24]: Changing these here isn't needed. Only the tests where a Http::fake was used should be replaced. The rest should still hit Paddle's API.

**Review by driesvints — CHANGES_REQUESTED** [review]

File: tests/Feature/WebhooksTest.php
Comment (kylemilloy) [reply tests/Feature/WebhooksTest.php:24]: So the way I have it set up is that `Cashier::fake()` fakes the events fired by Cashier (with the exception of the webhook events) as well as the Http calls. Confirming you would like me to remove the faked events from this and pull out all the assertions I added in as well.

File: .github/CONTRIBUTING.md
Comment (driesvints) [reply .github/CONTRIBUTING.md:11]: This lgtm 👍

File: tests/Feature/SubscriptionsTest.php
Comment (driesvints) [inline tests/Feature/SubscriptionsTest.php:210]: I think I'd like to explicitly not make use of variables for this and duplicate the checked data.

File: tests/Feature/WebhooksTest.php
Comment (driesvints) [reply tests/Feature/WebhooksTest.php:24]: I see what you did now and that's actually pretty neat. Let's keep that 👍

**Review by driesvints — CHANGES_REQUESTED** [review]
So let's pull out the `defaultVendorEndpoints` stuff and let people always define the "response" part.

File: src/CashierFake.php
Comment (kylemilloy) [reply src/CashierFake.php]: Anyways...I moved this to the constructor in CashierFake if you want to review...

File: src/CashierFake.php
Comment (driesvints) [inline src/CashierFake.php:33]: Do we still need these if they're always explicitly given now?

File: src/CashierFake.php
Comment (driesvints) [reply src/CashierFake.php]: Yes that's okay.

File: src/CashierFake.php
Comment (driesvints) [inline src/CashierFake.php:78]: Place this on one line.

File: tests/Feature/ChargesTest.php
Comment (driesvints) [inline tests/Feature/ChargesTest.php:39]: This can be removed.

**Review by driesvints — CHANGES_REQUESTED** [review]

**Comment by driesvints** [comment]
What if we moved `response` up another level?

```php
Cashier::fake([
    'subscription/modifiers' => [[
        'modifier_id' => 6789,
        'sucscription_id' => 3423423,
        'amount' => 15.00,
        'currency' => 'EUR',
        'is_recurring' => false,
        'description' => 'This is a test modifier',
    ]]
]);
```

I guess the only downside of this is that you can't set the success state per-endpoint anymore. But tbh that's okay I think.

File: src/CashierFake.php
Comment (kylemilloy) [reply src/CashierFake.php:33]: Ehhhhh...it depends. 

So the way I intended this `Cashier::fake()` to be used is that a dev could put in a single `Cashier::fake()` and it would automatically just mock any API endpoint and event that Cashier touched. The user doesn't really care about what values are returned from the API, just that the call was successful and Cashier wouldn't throw. If the user cared about what was returned here then they could override the values by providing their own data such as:

```php
Cashier::fake([
  'payment/refund' => [/* ... */]
]);
```

If we remove this list of endpoints from the constructor I think it limits the usefulness of the method because now everytime we're asking a dev to provide the endpoints that need to be mocked every time they want to do a fake. This asks the dev to do two things: 1. Know how Cashier works under the hood by knowing what methods call what endpoints, and 2. Knowing the Paddle API and what values need to be returned (ie, every response _needs_ to have `success: true`

So while we _could_ remove these endpoints and leave it up to the developer using `Cashier::fake()` to do it every time I think that eliminates the usefulness of this method.

I think some of that purpose is getting lost in the back and forth of the code review here so I'm willing to jump on a quick call to get this sorted. You're obviously not in my timezone (GMT-6) but I'm willing to make myself available to just quickly bang through this if you'd like instead. Feel free to DM me or I can ping you on twitter or something....lemme know what's best.

**Comment by kylemilloy** [comment]
> What if we moved response up another level?

It depends. I think it's important to be able to override the success in-case you want to fake an error and confirm that Cashier cancels a subscription or your app handles an error gracefully. 

I was thinking maybe there's some methods like `Cashier::fake()->error($message)` or `Cashier::fake()->endpoint($endpoint, $response)` where it would help a dev in crafting these things but then that feels outside the scope of what other `fake` methods in Laravel do...

**Comment by driesvints** [comment]
> I was thinking maybe there's some methods like Cashier::fake()->error($message) or Cashier::fake()->response($endpoint, $response) we could add in order to help a dev in crafting custom responses but I felt that that behaviour would diverge from what other fake methods in Laravel do...

I think that's a really good idea actually. That's a nice looking API. I'd still allow to accept the endpoint/responses as is in the current way but the fluent syntax looks good.

File: src/CashierFake.php
Comment (driesvints) [reply src/CashierFake.php:33]: You're correct, this is fine 👍

**Comment by kylemilloy** [comment]
@driesvints -- okay...added this in...via the constructor you can overwrite the entire response but it will default to `'success' => true` unless you change that. `response($endpoint, $response)` sets the given response to true, `error($endpoint, $message, $code)` sets the given endpoint to an error. 

Give it a look and let me know what you think.

**Comment by driesvints** [comment]
@kylemilloy thanks a lot for all your work on this. This turned out into a great PR 👍 

@taylorotwell this is ready to be reviewed now. Let me know if you need any questions.

**Comment by driesvints** [comment]
Tbh I don't know if we should take any action for the stripe side of things. You already have stripe mock which should provide anything you need.

**Comment by kylemilloy** [comment]
Totally...but on the flipside you could just use `Http::fake()` or `Event::fake()` for the paddle one which provides anything you need too. It's more for the improved dev experience where I simply say `Cashier::fake()` and it does the heavy lifting without me thinking about it that I like. Kinda like the rest of Laravel 😉 

Either way...this has to get approved first and then if that happens I'll look into the effort for something on the stripe side. I have another project that uses OG Cashier where I can bounce that idea around.

**Comment by kylemilloy** [comment]
@driesvints So I was playing around with Stripe mock and ya...you're totally right. It's just a way easier way of doing it than trying to break apart their SDK. 

That said...maybe this is a job for sail where we include it as one of the install options when setting up your server so it injects stripe mock into your docker-compose like:

```yml
stripemock:
  image: 'stripemock/stripe-mock:latest'
  networks:
    - sail
```

Probably take this discussion off of this thread but would be welcome to see feedback.

**Comment by driesvints** [comment]
@kylemilloy I don't think we're going to add that as a default to Sail as it just isn't of interest to 90% of Sail users.

**Comment by kylemilloy** [comment]
@driesvints no sweat.

**Comment by driesvints** [comment]
@kylemilloy wrote some docs for this but not entirely happy with the examples: https://github.com/laravel/docs/pull/7964

**Pull request merged**
